test/Validate.SSA.cpp

```diff
@@ -0,0 +1,699 @@
+}
+
+TEST_F(Validate, ForwardFunctionCallGood) {
+  char str[] = R"(
+        OpMemoryModel Logical GLSL450
+        OpEntryPoint GLCompute %5 ""
+        OpExecutionMode %5 LocalSize 1 1 1
+%1    = OpTypeVoid
+%2    = OpTypeInt 32 1
+%3    = OpTypeInt 32 0
+%4    = OpTypeFunction %1
+%5    = OpFunction %1 None %4
+%6    = OpLabel
+%four = OpConstant %2 4
+%five = OpConstant %3 5
+%7    = OpFunctionCall %1 %9 %four %five
+        OpFunctionEnd
+%8    = OpTypeFunction %1 %2 %3
+%9    = OpFunction %1 None %8
+%10   = OpFunctionParameter %2
+%11   = OpFunctionParameter %3
+%12   = OpLabel
+        OpReturn
+        OpFunctionEnd
+)";
+  ValidateInstructions(str, SPV_SUCCESS);
+}
+
+TEST_F(Validate, DISABLED_ForwardBranchConditionalGood) {
+  char str[] = R"(
+%voidt  = OpTypeVoid
+%boolt  = OpTypeBool
+%vfunct = OpTypeFunction %voidt
+%main   = OpFunction %voidt None %vfunct
+%true   = OpConstantTrue %boolt
+          OpSelectionMerge %endl None
+          OpBranchConditional %true %truel %falsel
+%truel  = OpLabel
+          OpNop
+          OpBranch %endl
+%falsel = OpLabel
+          OpNop
+%endl    = OpLabel
+          OpReturn
+          OpFunctionEnd
+)";
+  ValidateInstructions(str, SPV_SUCCESS);
+}
+
+TEST_F(Validate, DISABLED_ForwardBranchConditionalWithWeightsGood) {
+  char str[] = R"(
+%voidt  = OpTypeVoid
+%boolt  = OpTypeBool
+%vfunct = OpTypeFunction %voidt
+%main   = OpFunction %voidt None %vfunct
+%true   = OpConstantTrue %boolt
+          OpSelectionMerge %endl None
+          OpBranchConditional %true %truel %falsel !1 !9
+%truel  = OpLabel
+          OpNop
+          OpBranch %endl
+%falsel = OpLabel
+          OpNop
+%endl   = OpLabel
+          OpReturn
+          OpFunctionEnd
+)";
+  ValidateInstructions(str, SPV_SUCCESS);
+}
+
+TEST_F(Validate, ForwardBranchConditionalNonDominantConditionBad) {
+  char str[] = R"(
+%voidt  = OpTypeVoid
+%boolt  = OpTypeBool
+%vfunct = OpTypeFunction %voidt
+%main   = OpFunction %voidt None %vfunct
+          OpSelectionMerge %endl None
+          OpBranchConditional %true %missing %falsel
+%truel  = OpLabel
+          OpNop
+          OpBranch %endl
+%falsel = OpLabel
+          OpNop
+%endl   = OpLabel
+%true   = OpConstantTrue %boolt
+          OpReturn
+          OpFunctionEnd
+)";
+  ValidateInstructions(str, SPV_ERROR_INVALID_ID);
+}
+
+TEST_F(Validate, ForwardBranchConditionalMissingLabelBad) {
+  char str[] = R"(
+%voidt  = OpTypeVoid
+%boolt  = OpTypeBool
+%vfunct = OpTypeFunction %voidt
+%main   = OpFunction %voidt None %vfunct
+%true   = OpConstantTrue %boolt
+          OpSelectionMerge %endl None
+          OpBranchConditional %true %missing %falsel
+%truel  = OpLabel
+          OpNop
+          OpBranch %endl
+%falsel = OpLabel
+          OpNop
+%endl   = OpLabel
+          OpReturn
+          OpFunctionEnd
+)";
+  ValidateInstructions(str, SPV_ERROR_INVALID_ID);
+}
+
+const string kBasicTypes = R"(
+%voidt  = OpTypeVoid
+%boolt  = OpTypeBool
+%int8t  = OpTypeInt 8 0
```

Comment:
Universal validation rule says "Scalar integer types can be parameterized only as 32 bit, plus any additional sizes enabled by capabilities.
So any time you include "OpTypeInt 8  <anything>" you also need 
```OpCapability Int8``` right at the start of the module.